File: src/main/java/david/java/flink/watermark/TimeLagWatermarkGenerator.java
package david.java.flink.watermark;

import org.apache.flink.api.common.eventtime.Watermark;
import org.apache.flink.api.common.eventtime.WatermarkGenerator;
import org.apache.flink.api.common.eventtime.WatermarkOutput;

/**
 * @Description: a periodic watermark work on processing time
 * @Author: Zhaodawei
 * @Date: Create in 上午11:01 2020/12/21
 */
public class TimeLagWatermarkGenerator implements WatermarkGenerator {
    private final long maxTimeLag = 5000;

    @Override
    public void onEvent(Object o, long l, WatermarkOutput watermarkOutput) {
        // don't need to do anything because we work on peocessing time
    }

    @Override
    public void onPeriodicEmit(WatermarkOutput watermarkOutput) {
        watermarkOutput.emitWatermark(new Watermark(System.currentTimeMillis() - maxTimeLag));
    }
}

File: src/main/scala/david/scala/flink/watermark/TimeLagWatermarkGenerator.scala
package david.scala.flink.watermark

import david.java.flink.pojo.MyEvent
import org.apache.flink.api.common.eventtime.{Watermark, WatermarkGenerator, WatermarkOutput}

/**
 * @Description:
 * @Author: Zhaodawei
 * @Date: Create in 下午1:21 2020/12/21 
 */
class TimeLagWatermarkGenerator extends WatermarkGenerator[MyEvent]{

    val maxTimeLag: Long = 5000L

    override def onEvent(t: MyEvent, l: Long, watermarkOutput: WatermarkOutput): Unit = {
        // pass
    }

    override def onPeriodicEmit(watermarkOutput: WatermarkOutput): Unit = {
        watermarkOutput.emitWatermark(new Watermark(System.currentTimeMillis() - maxTimeLag))
    }
}
